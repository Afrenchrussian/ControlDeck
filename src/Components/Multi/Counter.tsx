import * as React from "react";
import { useState } from "react";
import { ButtonBase, Grid, makeStyles, Typography } from "@material-ui/core";
import { Add, Remove } from "@material-ui/icons";

const useStyles = makeStyles((theme) => ({
    root: {
        backgroundColor: theme.palette.primary.main,
        width: 70,
        height: 40
    },
    boxLeft: {
        lineHeight: "35px",
        textAlign: "center",
        width: "35px",
        height: "40px",
        borderRight: "solid 1px #707070"
    },
    boxRight: {
        lineHeight: "35px",
        textAlign: "center",
        width: "35px",
        height: "40px",
        borderLeft: "solid 1px #707070"
    },
    icons: {
        verticalAlign: "middle",
        color: "white",
        justifySelf: "center"
    },
    text: {
        color: "white",
        alignSelf: "center",
        textAlign: "right"
    }
}));

/*
    title: will be shown to the right of the counter //TESTED
    maxValue: highest value the counter can have, if not set default is 99 //TESTED
    minValue: minimum value the counter can have, if not set 0 is default //TESTED
    defaultValue: the initial value of the counter on load //TESTED
    onChange: a function that will be ran on each change and will pass back the current value //TESTED
 */

interface Props {
    title: string,
    maxValue?: number,
    minValue?: number,
    defaultValue?: number
    onChange?: (value: number) => void
}

export const Counter = (props: Props) => {
    const [counter, setCounter] = useState((props.defaultValue === undefined) ? 0 : props.defaultValue);
    const classes = useStyles();
    return (
        <Grid container direction={"row"} style={{ marginBottom: "10px", width: "100%" }} justify={"center"}>
            <Grid item className={classes.text} style={{ marginRight: "10px", width: "70px" }}>
                <Typography id={"Counter_Text"}>
                    {props.title}
                </Typography>
            </Grid>
            <Grid item className={classes.root}>
                <ButtonBase id={"Counter_DecreaseButton"} onClick={() => {
                    if (counter - 1 >= (props.minValue || 0)) {
                        setCounter(counter - 1);
                        if (props.onChange === undefined) return;
                        props.onChange(counter - 1);
                    }
                }}>
                    <Grid item className={classes.boxLeft}>
                        <Remove className={classes.icons}/>
                    </Grid>
                </ButtonBase>
                <ButtonBase id={"Counter_IncreaseButton"} onClick={() => {
                    if (counter >= (props.maxValue || 99)) return;
                    setCounter(counter + 1);
                    if (props.onChange === undefined) return;
                    props.onChange(counter + 1);
                }}>
                    <Grid item className={classes.boxRight}>
                        <Add className={classes.icons}/>
                    </Grid>
                </ButtonBase>
            </Grid>
            <Grid item className={classes.text} style={{ marginLeft: "10px", width: "20px" }}>
                <Typography id={"Counter_Value"}>
                    {counter}
                </Typography>
            </Grid>
        </Grid>
    );
};
