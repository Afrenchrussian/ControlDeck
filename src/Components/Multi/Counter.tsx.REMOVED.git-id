8bc611ab1cf8a4507955f46105c628ad4d00988a