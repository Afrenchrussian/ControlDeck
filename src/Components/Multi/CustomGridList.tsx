import React from "react";
import { GridList, GridListTile, makeStyles, Tooltip, Typography } from "@material-ui/core";
import ButtonBase from "@material-ui/core/ButtonBase";
import { ErrorOutlineTwoTone, PhoneIphone, TabletMac } from "@material-ui/icons";
import { Device, Profile } from "../../Store/Reducers/mainReducer";

/*
    data: can be any of the types listed as only a name and a type is required
          (name : used for id and shown in tile // TESTED , type: used to pick which icon to show )
    onTileClick: when a tile is clicked this call back is called with the index of the tile clicked // TESTED
    checkForWarnings: if defined should return a string containing a warning for any cases, // TESTED
                        callback will be passed the elements data
    //TODO deletable: NOT IMPLEMENTED -- when true will allow any of the tiles to be deleted
    //TODO onDelete: NOT IMPLEMENTED -- called when a tile is clicked to be deleted
 */

interface Props {
    data: Profile[] | Device[] | {name: string, type: number}[],
    onTileClick?: (index: number) => void
    checkForWarnings?: (element: any) => string,
    deletable?: boolean,
    onDelete?: (index: number) => void
}

const useStyles = makeStyles((theme) => ({
    card: {
        width: "180px",
        height: "110px",
        backgroundColor: theme.palette.primary.main,
        "&:hover": {
            boxShadow: "0px 3px 6px #00000029",
            cursor: "pointer"
        }
    },
    text: {
        margin: "auto",
        textTransform: "capitalize",
        color: "white"
    },
    icon_large: {
        fontSize: "55px",
        color: "white",
        margin: "auto"
    },
    icon_small: {
        fontSize: "35px",
        color: "white",
        margin: "auto"
    }
}));

export default function CustomGridList(props: Props) {
    const classes = useStyles();

    const icon = (type: number) => {
        switch (type) {
            case 0:
                return <PhoneIphone className={classes.icon_large}/>;
            case 1:
                return <PhoneIphone className={classes.icon_small}/>;
            case 2:
                return <TabletMac className={classes.icon_large}/>;
            case 3:
                return <TabletMac className={classes.icon_small}/>;
        }
    };

    return (
        <GridList cols={3} style={{ width: "602px", margin: "auto", marginTop: "30px" }}>
            {
                (props.data as {name: string, type: number}[]).map((cur, index) => {
                const warning = (props.checkForWarnings !== undefined && props.checkForWarnings(cur) !== "") ? (
                    <Tooltip id={cur.name + "_warning"} title={props.checkForWarnings(cur)}>
                        <ErrorOutlineTwoTone
                            style={{ color: "yellow", position: "absolute", opacity: 0.6, right: "5px", top: "5px" }}/>
                    </Tooltip>
                ) : <React.Fragment/>;

                return (
                    <GridListTile cols={1} key={index} style={{ height: "135px" }}>
                        <ButtonBase id={cur.name} onClick={() => {
                            if (props.onTileClick === undefined) return;
                            props.onTileClick(index);
                        }} style={{
                            width: "180px",
                            height: "110px"
                        }}>
                            <div className={classes.card}
                                 style={((index + 2) % 3 === 2) ? { marginLeft: 0 } : ((index + 2) % 3 === 1) ? {
                                     marginRight: 0,
                                     float: "right"
                                 } : { margin: "auto" }}>
                                {warning}
                                <div id={`${cur.name}_list_icon`} style={{ width: "100%", height: "80%", display: "grid" }}>
                                    {icon(cur.type)}
                                </div>
                                <Typography id={`${cur.name}_text`} className={classes.text}>
                                    {cur.name}
                                </Typography>
                            </div>
                        </ButtonBase>
                    </GridListTile>
                );
            })}
        </GridList>
    );
}
