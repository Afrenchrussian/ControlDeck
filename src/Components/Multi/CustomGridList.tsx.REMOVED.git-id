7ce9c5d0562932f42c19a47b407d0d3d166b8f60