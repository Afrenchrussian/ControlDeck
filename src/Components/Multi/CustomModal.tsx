import * as React from "react";
import { Grid, makeStyles, Modal, Typography } from "@material-ui/core";
import { CloseOutlined } from "@material-ui/icons";

/*
    title: shown at the top of the modal
    open: whether the modal should be shown or not
    onClose: a function ran when the use clicks the exit button
    children: elements that should be shown in the modal
    height: the height of the modal
    width: the width of the modal
 */

interface Props {
    title: string
    open: boolean,
    onClose: () => void,
    children: JSX.Element[] | JSX.Element
    height?: number,
    width?: number,
}

const useStyles = makeStyles((theme) => ({
    modal: {
        "&:focus": {
            outline: "none"
        }
    },
    container: {
        position: "relative",
        "&:focus": {
            outline: "none"
        },
        height: "100%"
    },
    root: (props: { height: number, width: number }) => ({
        height: `${props.height}px`,
        width: `${props.width}px`,
        backgroundColor: "#442354",
        position: "absolute",
        top: "50%",
        left: "50%",
        transform: "translate(-50%,-50%)"
    }),
    exitButton: {
        position: "absolute",
        top: "2px",
        right: "3px",
        color: "white",
        "&:hover": {
            cursor: "pointer"
        }
    },
    topBar: {
        textAlign: "center",
        height: "30px",
        width: "100%",
        backgroundColor: "#1B1B1B"
    },
    textBlock: {
        color: "white",
        fontSize: "13px",
        fontWeight: "bold",
        paddingTop: "15px",
        paddingLeft: "12px",
        paddingRight: "12px"
    },
    title: {
        lineHeight: "1.9", color: "white", fontWeight: "bold"
    }
}));

export const CustomModal = (props: Props) => {
    const classes = useStyles({
        height: (props.height !== undefined) ? props.height : 380,
        width: (props.width !== undefined) ? props.width : 290
    });
    return (
        <Modal
            id={`${props.title}_modal`}
            className={classes.modal}
            open={props.open}
        >
            <div className={classes.container}>
                <Grid container direction={"column"} className={classes.root}>
                    <div className={classes.topBar}>
                        <Typography id={`${props.title}_title`} className={classes.title}>{props.title}</Typography>
                        <CloseOutlined id={`${props.title}_close_button`} className={classes.exitButton} onClick={() => props.onClose()}/>
                    </div>
                    {props.children}
                </Grid>
            </div>
        </Modal>
    );
};
