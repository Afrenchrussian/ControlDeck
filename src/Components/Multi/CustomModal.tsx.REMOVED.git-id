bfe811803872948aaf6a72c157c75cdd1fd051d9