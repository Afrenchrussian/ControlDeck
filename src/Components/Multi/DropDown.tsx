import React, { useState } from "react";
import { FormControl, InputLabel, makeStyles, MenuItem, Select } from "@material-ui/core";

const useStyles = makeStyles((theme) => ({
    root: {
        color: "white",
        width: "100%"
    },
    label: {
        color: "white",
        fontSize: "14px",
        padding: "4px 0 0 5px"
    },
    labelFocused: {
        color: "white !important"
    },
    textBlock: {
        color: "white",
        fontSize: "15px"
    },
    selectPaper: {
        marginTop: "40px"
    },
    selectedElement: {
        backgroundColor: "#693680",
        color: "white",
        fontSize: "15px",
        paddingLeft: "10px",
        paddingTop: "6px"
    },
    select: {
        left: "50%",
        transform: "translateX(-50%)",
        "& .MuiInput-underline:before": {
            borderBottom: "1px solid white"
        },
        "& .MuiInput-underline:after": {
            borderBottom: "1px solid white"
        }
    }

}));

export interface AdvancedElem {
    value: any,
    name: string
}

interface Props {
    title?: string
    elements: string[] | number[] | AdvancedElem[],
    onChange: ((elem: any) => void),
    defaultValue?: any
}

export default function DropDown(props: Props) {
    const classes = useStyles();
    const [value, setValue] = useState(props.defaultValue || (typeof (props.elements[0]) === "string" || typeof (props.elements[0]) === "number") ? props.elements[0] : 0);

    const handleChange = (event: React.ChangeEvent<{ value: unknown }>) => {
        setValue(event.target.value as string);
        props.onChange(event.target.value as string);
    };

    const mainSelect = (
        <Select
            labelId="demo-customized-select-label"
            id="demo-simple-select"
            MenuProps={{ classes: { paper: classes.selectPaper } }}
            inputProps={{ classes: { root: classes.textBlock } }}
            className={classes.select}
            value={value}
            fullWidth
            onChange={handleChange}
        >
            {
                (typeof (props.elements[0]) === "string" || typeof (props.elements[0]) === "number") ?
                    (props.elements as Array<string | number>).map((cur: string | number) => (
                        <MenuItem key={cur} value={cur}>{cur}</MenuItem>
                    )) :
                    (props.elements as AdvancedElem[]).map((cur: AdvancedElem, index: number) => (
                        <MenuItem key={index} value={cur.value}>{cur.name}</MenuItem>
                    ))

            }
        </Select>
    );

    return (props.title !== undefined) ? (
        <FormControl fullWidth>
            <InputLabel className={classes.label} classes={{ focused: classes.labelFocused }}
                        id="demo-simple-select-label">{props.title}</InputLabel>
            {mainSelect}
        </FormControl>
    ) : (
        <React.Fragment>
            {mainSelect}
        </React.Fragment>
    );
}
