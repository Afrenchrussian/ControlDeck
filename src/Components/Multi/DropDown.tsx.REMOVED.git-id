980236414bc98a04a487215e67c1c3efb7cc7af9