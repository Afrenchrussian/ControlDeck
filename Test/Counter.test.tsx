import React from "react";
import renderer from "react-test-renderer";
import { Counter } from "../src/Components/Multi/Counter";
import { shallow } from "enzyme";

test("Counter renders correctly", () => {
    const counter = renderer.create(<Counter title={"Test"} defaultValue={0}/>);
    const counterTree = counter.toJSON();
    expect(counterTree).toMatchSnapshot();
});

test("Counter title is correct", () => {
    const counter = shallow(<Counter title={"Test_Name"} defaultValue={0}/>);
    expect(counter.find("#Counter_Text").text()).toEqual("Test_Name");
});

test("Increase counter", () => {
    const counter = shallow(<Counter title={"Test_Name"} defaultValue={0}/>);
    counter.find("#Counter_IncreaseButton").simulate("click");
    expect(counter.find("#Counter_Value").text()).toEqual("1");
});

test("Decrease counter", () => {
    const counter = shallow(<Counter title={"Test_Name"} defaultValue={2}/>);
    counter.find("#Counter_DecreaseButton").simulate("click");
    expect(counter.find("#Counter_Value").text()).toEqual("1");
});

test("Test min value prop", () => {
    const counter = shallow(<Counter title={"Test_Name"} defaultValue={2} minValue={1}/>);
    expect(counter.find("#Counter_Value").text()).toEqual("2");
    for (let i = 3; i > 0; i--) {
        counter.find("#Counter_DecreaseButton").simulate("click");
        expect(counter.find("#Counter_Value").text()).toEqual("1");
    }
});

test("Test max value prop", () => {
    const counter = shallow(<Counter title={"Test_Name"} defaultValue={0} maxValue={2}/>);
    for (let i = 0; i <= 2; i++) {
        expect(counter.find("#Counter_Value").text()).toEqual(`${i}`);
        counter.find("#Counter_IncreaseButton").simulate("click");
    }
});

test("OnChange with Increase", () => {
    const mockCallback = jest.fn(value => {
    });
    const counter = shallow(<Counter title={"Test_Name"} defaultValue={0} onChange={mockCallback}/>);
    for (let i = 1; i < 3; i++) {
        counter.find("#Counter_IncreaseButton").simulate("click");
        expect(mockCallback.mock.calls.length).toBe(i);
        expect(mockCallback).toHaveBeenCalledWith(i);
    }
});

test("OnChange with Decrease", () => {
    const mockCallback = jest.fn((value) => value);
    const counter = shallow(<Counter title={"Test_Name"} defaultValue={3} onChange={mockCallback}/>);
    for (let i = 3; i > 0; i--) {
        counter.find("#Counter_DecreaseButton").simulate("click");
        expect(mockCallback.mock.calls.length).toBe(4 - i);
        expect(mockCallback).toHaveBeenCalledWith(i - 1);
    }
});
