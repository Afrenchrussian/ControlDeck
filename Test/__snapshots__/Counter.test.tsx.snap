// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter renders correctly 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-center"
  style={
    Object {
      "marginBottom": "10px",
      "width": "100%",
    }
  }
>
  <div
    className="MuiGrid-root makeStyles-text-5 MuiGrid-item"
    style={
      Object {
        "marginRight": "10px",
        "width": "70px",
      }
    }
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
      id="Counter_Text"
    >
      Test
    </p>
  </div>
  <div
    className="MuiGrid-root makeStyles-root-1 MuiGrid-item"
  >
    <button
      className="MuiButtonBase-root"
      disabled={false}
      id="Counter_DecreaseButton"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <div
        className="MuiGrid-root makeStyles-boxLeft-2 MuiGrid-item"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root makeStyles-icons-4"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13H5v-2h14v2z"
          />
        </svg>
      </div>
    </button>
    <button
      className="MuiButtonBase-root"
      disabled={false}
      id="Counter_IncreaseButton"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <div
        className="MuiGrid-root makeStyles-boxRight-3 MuiGrid-item"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root makeStyles-icons-4"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
      </div>
    </button>
  </div>
  <div
    className="MuiGrid-root makeStyles-text-5 MuiGrid-item"
    style={
      Object {
        "marginLeft": "10px",
        "width": "20px",
      }
    }
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
      id="Counter_Value"
    >
      0
    </p>
  </div>
</div>
`;
